### ISSUE TITLE
딥러닝 입문 기말 과제 : CNN을 통한 얼굴 감정 인식 모델

### ISSUE THREAD

**Issue by SihwanGit** [SihwanGit/DeepLearningIntroduction#4]
과제의 요건은 아래와 같다. 
 
1. 과제 목표Kaggle 얼굴 표정 데이터셋을 활용하여 CNN 기반 이미지 분류 모델을 구현하고, 테스트 데이터에 대한 예측 정확도를 최대화하는 것이 목표본 과제는 총 20점 만점
2. 과제 개요
데이터셋: https://www.kaggle.com/datasets/jonathanoheix/face-expression-recognition-dataset
입력 X: 48×48 픽셀 grayscale 얼굴 이미지
출력 Y: 7가지 감정 클래스
angry, disgust, fear, happy, neutral, sad, surprise
목표: 테스트 데이터에 대한 분류 정확도 최대화
3. 개발 및 사용 가능 모듈
개발 환경: Google Colab 주피터 노트북 필수, T4 GPU 런타임
사용 가능 파이썬 모듈: torch, torchvision, kagglehub, pandas, numpy, matplotlib.pyplot, PIL(Pillow), os, glob, time, datetime 및 각 하위 모듈
4. 요구사항
입력: -1 ~ 1 정규화 권장, 출력은 전처리 하지 않음. 상기 모듈 및 파이썬 범위 내에서 자유롭게 정규화 가능
학습시간은 20분 이내
TensorBoard로 학습 손실 출력 구현 필수
마크다운: 최종 결과뿐만 아니라, 본인이 노력한 과정도 기술
CNN 모델 구현 필수
Convolution, Pooling, Fully Connected Layer 등 자유롭게 조합
CNN 구조(층 수, 커널 크기, 채널 수 등)는 자유롭게 설계
CNN 범위를 벗어나는 Transformer, LSTM 등 시퀀스 모델 사용 불가
5. 필수 구현 사항
1) 시간 출력
학습하는 셀의 처음과 마지막에서 시간 측정 및 출력
2) TensorBoard 시각화
TensorBoard를 이용하여 학습 과정의 손실(loss) 및 정확도(accuracy)를 기록 및 시각화 할 것
관련 로그를 남기는 코드 필수
3) 최종 성능 평가
학습 완료 후, 테스트셋에 대한 최종 정확도(Accuracy) 수치를 출력할 것
과제 미제출자는 정확도 0으로 처리
4) 마크다운 기술
코드만 제출하지 말고, 본인이 시도한 과정과 내용을 마크다운으로 정리할 것
필수 포함 내용 예시:
사용한 CNN 구조 설명
하이퍼파라미터 선택 이유 (learning rate, batch size, epoch 수 등)
실험 결과 및 분석 (정확도 비교, 실패/성공 시도 등)
수업 시간에 다루지 않은 내용을 사용한 경우, 해당 부분에 대해 마크다운으로 반드시 설명을 덧붙일 것
6. 제출 방식
마감 기한: 12/4(목) 18:00
마감 이후 추가 제출 및 수정 불가
제출 파일
Colab 노트북 파일(.ipynb) 1개
모든 셀을 위에서부터 순서대로 실행 가능해야 함
파일명 형식: 이름_학번.ipynb
예: 홍길동_2025123456.ipynb
별도의 문서(hwp, docx, pdf 등)는 제출하지 않음
기타 유의사항
실행 오류가 발생하지 않도록 사전에 반드시 확인
데이터 경로, Kaggle 연동 등이 코랩 내에서 재현 가능해야 함
7. 평가 기준 (총 20점)
1. 정확도 (10점)
제출한 ipynb파일을 심사시에 재실행 한 후 정확도 기준
정확도에 따라 점수 부여 예정
학습시간 20분 초과시 5점 감점, 30분 초과시 0점 처리
2. 코드 가독성 (3점)
변수명, 함수명, 코드 구조, 주석 여부 등을 종합적으로 평가
3. 마크다운 설명 (3점)
모델 설계 의도, 실험 과정, 결과 해석 등이 얼마나 잘 정리되어 있는지 평가
4. 상위 3명 과제 공유 (3점)
상위 성능 3명은 코드 공유 진행
5. 기본 점수 (1점)
제출한 학생에게 공통 부여
8. 예시코드

[시간 출력 코드]
학습 이전 부분: 
start_time = time.time()
print(f"학습 시작: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")

학습 이후 부분
end_time = time.time()
elapsed_time = (end_time - start_time) /60

print(f"학습 종료: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
print(f"소요 시간: {elapsed_time:.2f}분")

[입력 데이터셋 로드 예시]
# 데이터셋 다운로드
path = kagglehub.dataset_download("jonathanoheix/face-expression-recognition-dataset")
train_dir = os.path.join(path, 'images', 'train')
test_dir = os.path.join(path, 'images', 'validation')

# 기본 변환 (48×48, grayscale)
transform = transforms.Compose([
transforms.Resize((48, 48)), # 필수: 48×48 크기
transforms.Grayscale(num_output_channels=1), # 필수: grayscale 변환
transforms.ToTensor(),
transforms.Normalize((0.5,), (0.5,)) # -1 ~ 1 정규화 (권장)
])

# 데이터셋 로드
train_dataset = datasets.ImageFolder(train_dir, transform=transform)
test_dataset = datasets.ImageFolder(test_dir, transform=transform)

[정확도 출력 예시]
# 함수명과 변수명은 자유롭게
model.eval()
correct = 0
total = 0

with torch.no_grad():
for images, labels in test_loader:
images, labels = images.to(device), labels.to(device)
outputs = model(images)
_, predicted = torch.max(outputs, 1)
correct += (predicted == labels).sum().item()
total += labels.size(0)

test_accuracy = 100 * correct / total
print(f"정확도: {test_accuracy:.2f}%")
